### ISSUE TITLE
Git-2.40.0-32-bit.exe - Pop-up message in silent installation

### ISSUE THREAD

**Issue by metaad** [git-for-windows/git#4347]
32-bit installer (Git-2.40.0-32-bit.exe) throws pop-up while carrying out silent installation. 
The command line used for silent installation is 

**Git-2.40.0-32-bit.exe /SP- SUPPRESSMSGBOXES /VERYSILENT**

![image](https://user-images.githubusercontent.com/106653425/225519716-6d2c1df4-222e-4897-942a-4973f7c09d30.png)

**Can we get this pop-up removed while installing in silent mode?**

**Comment by rimrul** [comment]
>  32-bit installer (Git-2.40.0-32-bit.exe) throws pop-up while carrying out silent installation. 
> The command line used for silent installation is 

That message is already a `SuppressibleMsgBox`, so it can already be suppressed with `/SUPPRESSMSGBOXES`

>
> **Git-2.40.0-32-bit.exe /SP- SUPPRESSMSGBOXES /VERYSILENT**

That seems to be missing a `/`.

**Comment by metaad** [comment]
I have tried with **Git-2.40.0-32-bit.exe /SP- /SUPPRESSMSGBOXES /VERYSILENT.**  This does not install the application.
Also I tried installing in session 0 (Attached screenshot) , it completes with error code 1. 

![image](https://user-images.githubusercontent.com/106653425/225527924-004c57d1-4a34-4110-b469-05f6ca4e68e6.png)

**Comment by dscho** [comment]
> It is recommended to install the 64-bit variant of Git for Windows instead.

This message [is only shown on a 64-bit Windows](https://github.com/git-for-windows/build-extra/blob/a4798546318d6e6cee74e93cdb8fd446b91209bc/installer/install.iss#L1094-L1097). Please install the 64-bit variant instead.

**Comment by rimrul** [comment]
Oh. I think I see what's happening. The default choice for that MessageBox is No and `/SUPPRESSMSGBOXES` pretends the user chose the default. That's not ideal.

**Comment by dscho** [comment]
> Oh. I think I see what's happening. The default choice for that MessageBox is No and `/SUPPRESSMSGBOXES` pretends the user chose the default. That's not ideal.

@rimrul it is intentional, though. There is _no other way_ to tell users that they're using a 32-bit Git for Windows on a 64-bit Windows and that they need to switch. If the default would be `IDYES`, then everybody would ignore this message.

@metaad if you feel that there is a legitimate reason why you should prefer the 32-bit variant of Git for Windows instead, even if you can use the 64-bit variant, and even if the 32-bit variant is being phased out, you will want to imitate https://github.com/git-for-windows/build-extra/commit/ae32f65ea327547d91159f1c15e8255659b4e585 to force the installation to continue. To do that:

1. install [Git for Windows' SDK](https://gitforwindows.org/#download-sdk),
2. `sdk cd installer`,
3. edit `install.iss`

   something like this should get you started:

   ```diff
   diff --git a/installer/install.iss b/installer/install.iss
   index fd3d55ddc..5ba704307 100644
   --- a/installer/install.iss
   +++ b/installer/install.iss
   @@ -1086,15 +1086,21 @@ end;
    function InitializeSetup:Boolean;
    var
        CurrentVersion,Msg:String;
   -    ErrorCode:Integer;
   +    ErrorCode,Force32BitInstallDefault:Integer;
    begin
        UpdateInfFilenames;
    #if BITNESS=='32'
        Result:=True;
        if not IsWin64 then
            SuppressibleMsgBox('Git for Windows (32-bit) is nearing its end of support.'+#13+'More information at https://gitforwindows.org/32-bit.html',mbError,MB_OK or MB_DEFBUTTON1,IDOK)
   -    else if SuppressibleMsgBox('Git for Windows (32-bit) is nearing its end of support. It is recommended to install the 64-bit variant of Git for Windows instead.'+#13+'More information at https://gitforwindows.org/32-bit.html'+#13+'Continue to install the 32-bit variant?',mbError,MB_YESNO or MB_DEFBUTTON2,IDNO)=IDNO then
   +    else begin
   +        if ParamIsSet('FORCE32BITINSTALL`) then
   +            Force32BitInstallDefault:=IDYES
   +        else
   +            Force32BitInstallDefault:=IDNO;
   +        if SuppressibleMsgBox('Git for Windows (32-bit) is nearing its end of support. It is recommended to install the 64-bit variant of Git for Windows instead.'+#13+'More information at https://gitforwindows.org/32-bit.html'+#13+'Continue to install the 32-bit variant?',mbError,MB_YESNO or MB_DEFBUTTON2,Force32BitInstallDefault)=IDNO then
            Result:=False;
   +    end;
    #else
        if not IsWin64 then begin
            LogError('The 64-bit version of Git requires a 64-bit Windows. Aborting.');
   ```
4. build a new installer via `sdk build installer`/test the installer
5. verify that that installer works as expected?
6. open a PR?

**Comment by rimrul** [comment]
> > Oh. I think I see what's happening. The default choice for that MessageBox is No and `/SUPPRESSMSGBOXES` pretends the user chose the default. That's not ideal.
> 
> @rimrul it is intentional, though. There is _no other way_ to tell users that they're using a 32-bit Git for Windows on a 64-bit Windows and that they need to switch. If the default would be `IDYES`, then everybody would ignore this message.

Yes. I hadn't at that point realised that this version of the message was only shown on 64bit systems and not on 32bit Systems.

**Comment by jeremyd2019** [comment]
I get a similar message box in 2.40.0 on Windows 10 32-bit, and also on Windows 10 ARM64 (which cannot run x64 binaries, only x86).  On the latter I was attempting to do a silent installation and got the message box but no other UI, and would like to be able to suppress it and complete the installation silently again.

**Comment by jeremyd2019** [comment]
On win10 ARM64, 32-bit git for windows shows the yes/no variant of the message box, which defaults to no and results in `/SUPPRESSMSGBOXES` canceling the install rather than continuing.

**Comment by dscho** [comment]
> On win10 ARM64, 32-bit git for windows shows the yes/no variant of the message box, which defaults to no and results in `/SUPPRESSMSGBOXES` canceling the install rather than continuing.

I agree that this is rather unfortunate a behavior. I do lack the time to take care of this, though. Maybe you can help @jeremyd2019?

The code in question lives [here](https://github.com/git-for-windows/build-extra/blob/main/installer/install.iss#L1094-L1097). Sadly, I do not see any InnoSetup function [to tell whether the architecture is ARM32](https://jrsoftware.org/ishelp/topic_isxfunc_processorarchitecture.htm). Maybe you can spot a function I missed? But I guess that we could use [the `WizardSilent` function](https://jrsoftware.org/ishelp/topic_isxfunc_wizardsilent.htm) to determine whether to continue because the user has no chance to see the commit message? Or maybe we need something similar to [the `/ALLOWDOWNGRADE` option](https://github.com/git-for-windows/build-extra/commit/ae32f65ea327547d91159f1c15e8255659b4e585) here?

**Comment by rimrul** [comment]
> > On win10 ARM64, 
> […]
> Sadly, I do not see any InnoSetup function [to tell whether the architecture is ARM32](https://jrsoftware.org/ishelp/topic_isxfunc_processorarchitecture.htm).

The comment you're quoting is about ARM64, not ARM32.
This should be fixed by https://github.com/git-for-windows/build-extra/pull/492.

**Comment by jeremyd2019** [comment]
> The comment you're quoting is about ARM64, not ARM32. This should be fixed by [git-for-windows/build-extra#492](https://github.com/git-for-windows/build-extra/pull/492).

I did an install of 2.41.0 32-bit on Windows 10 ARM64, and got the message box that only has "OK", so it does indeed seem to address my concern, thanks.
